src/fragmenter.c

```diff
@@ -0,0 +1,412 @@
+#include "fragmenter.h"
+#include "toksep.h"
+#include "tokenize.h"
+#include <ctype.h>
+#include <float.h>
+#include <sys/uio.h>
+#include <assert.h>
+
+#define myMin(a, b) (a) < (b) ? (a) : (b)
+#define myMax(a, b) (a) > (b) ? (a) : (b)
+
+#ifdef __APPLE__
+#define myQsort_r(arr, nelem, elemSize, compfn, arg) qsort_r(arr, nelem, elemSize, arg, compfn)
+#define MY_QSORTR_ARGS(ctx, a, b) void *ctx, const void *a, const void *b
+#else
+#define myQsort_r(arr, nelem, elemSize, compfn, arg) qsort_r(arr, nelem, elemSize, compfn, arg)
+#define MY_QSORTR_ARGS(ctx, a, b) const void *a, const void *b, void *ctx
+#endif
+
+// Estimated characters per token
+#define EST_CHARS_PER_TOK 6
+
+static Fragment *FragmentList_LastFragment(FragmentList *fragList) {
+  if (!fragList->frags.offset) {
+    return NULL;
+  }
+  return (Fragment *)(fragList->frags.data + (fragList->frags.offset - sizeof(Fragment)));
+}
+
+static Fragment *FragmentList_AddFragment(FragmentList *fragList) {
+  Fragment *frag = Buffer_AddToSize(&fragList->frags, sizeof(Fragment));
+  memset(frag, 0, sizeof(*frag));
+  frag->fragPos = fragList->numFrags++;
+  return frag;
+}
+
+static size_t Fragment_GetNumTerms(const Fragment *frag) {
+  return BUFFER_GETSIZE_AS(&frag->termLocs, TermLoc);
+}
+
+static TermLoc *Fragment_AddTerm(Fragment *frag) {
+  TermLoc *loc = Buffer_AddToSize(&frag->termLocs, sizeof(TermLoc));
+  // memset(loc, 0xff, sizeof *loc);
+  return loc;
+}
+
+static int tokenizerCallback(void *ctx, const Token *tokInfo) {
+  FragmentList *fragList = ctx;
+  const FragmentTerm *term = NULL;
+  uint32_t termId;
+  int isStem = 0;
+
+  // See if this token matches any of our terms.
+  for (termId = 0; termId < fragList->numTerms; ++termId) {
+    const FragmentTerm *cur = fragList->terms + termId;
+    if (tokInfo->tokLen == cur->len && strncmp(tokInfo->tok, cur->tok, cur->len) == 0) {
+    } else if (tokInfo->stem && tokInfo->stemLen == cur->len &&
+               strncmp(tokInfo->stem, cur->tok, cur->len) == 0) {
+      isStem = 1;
```

Comment:
TODO: isStem is not actually used anywhere for calculations and such. Should we care if it's a stem or not?